docs/mitigating_ret2spec.md

```diff
@@ -0,0 +1,122 @@
+
+## Linux mitigations for ret2spec
+
+This discussion only covers mitigations on `x86`.
+
+There are four places where we need to worry about code consuming RSB entries created in another context:
+
+1. User process to user process
+2. User process to kernel
+3. Virtual machine to virtual machine
+4. Virtual machine to hypervisor
+
+The most straightforward mitigation for ret2spec is to fill or "stuff" the RSB with benign entries. The Linux kernel has a [`FILL_RETURN_BUFFER` assembler macro](https://github.com/torvalds/linux/blob/08bf1a27c4c354b853fd81a79e953525bbcc8506/arch/x86/include/asm/nospec-branch.h#L105) that accomplishes this, which is used on [context switches](https://github.com/torvalds/linux/blob/2646fb032f511862312ec8eb7f774aaededf310d/arch/x86/entry/entry_64.S#L253) and VM exits ([Intel VT-x](https://github.com/torvalds/linux/blob/921d2597abfc05e303f08baa6ead8f9ab8a723e1/arch/x86/kvm/vmx/vmenter.S#L83), [AMD SVM](https://github.com/torvalds/linux/blob/921d2597abfc05e303f08baa6ead8f9ab8a723e1/arch/x86/kvm/svm/vmenter.S#L107)).
+
+Filling the RSB as part of every context switch prevents user processes from attacking each other. The RSB fill immediately after VM exit prevents that VM from attacking other VMs or the hypervisor.
+
+This leaves one last attack vector: user to kernel. The Linux kernel does **not** fill the RSB on kernel entry. So, how is the kernel protected from user processes?
+
+## An appealing but incomplete answer
+
+A ret2spec attack is only possible when code consumes an RSB entry it didn't create -- when it executes a `RET` without an earlier paired `CALL`. Maybe that can't happen when we enter the kernel?
+
+Context switches and VM exits cause execution to resume at an arbitrary depth in a reactivated call stack, which means we'll encounter unpaired `RET`s and we need to worry about ret2spec. But when we enter the kernel to handle an interrupt or system call, code starts executing at a known entry point. It might seem like `CALL`s and `RET`s should all be properly paired after that -- at least, until we exit the kernel or hit a context switch.
+
+This doesn't **necessarily** hold true. For example, an especially clever system call handler might decide to implement a `CALL` as a `PUSH` and `JMP` -- maybe to cause the function to return somewhere other than the syscall handler. The resulting code works architecturally, but might speculatively consume an RSB entry populated by the user process.
+
+Somewhat less artificially, it is possible for an unpaired `RET` to be reached _speculatively_. Consider a system call implementation that chooses a "back half" handler and then jumps to it:
+
+```
```

Comment:
Can you make this an asm block?

~~~
```asm

...

```
~~~

This renders as:

```asm
; Given an operation index, look up the handler in an array and run it
SyscallHander(int):
 movsxd rax,edi                     ; rax = first argument
 jmp    QWORD PTR [rax*8+0x404040]  ; jump to entry in handler array
; Some function that happens to follow in memory and ends in `ret`
AnotherFunction():
 mov    eax,0x5
 ret
```

Reply:
Good change, done.